# SQL Проект

## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Загрузка библиотек

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Исследование таблиц

### Книги

In [2]:
query = '''
SELECT *
FROM books
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Авторы

In [3]:
query = '''
SELECT *
FROM authors
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### Издательства

In [4]:
query = '''
SELECT *
FROM publishers
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Рейтинг

In [5]:
query = '''
SELECT *
FROM ratings

'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

### Обзоры

In [6]:
query = '''
SELECT *
FROM reviews

'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [ ]:
query = '''
SELECT *
FROM reviews, ratings

'''
pd.io.sql.read_sql(query, con = engine)

## Задания

### Задание 1.

Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
query = '''
SELECT count(book_id) AS "Количество книг после 1 января 2000 года"
FROM books
Where publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(query, con = engine)

Количество книг после 1 января 2000 г
0                                    819

### Вывод

Количество книг, выпущенных после 1 января 2000 года - 819

### Задание 2.

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [8]:
query = '''

SELECT books.book_id AS "Книга",
       count (review_id) AS "Количество обзоров",
       AVG(rating) AS "Средняя оценка"

FROM books
LEFT JOIN reviews ON books.book_id = reviews.book_id 
LEFT JOIN ratings ON books.book_id = ratings.book_id
GROUP BY books.book_id
ORDER BY "Книга" ASC
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

Книга  Количество обзоров  Средняя оценка
0      1                   6        3.666667
1      2                   2        2.500000
2      3                   9        4.666667
3      4                   4        4.500000
4      5                  24        4.000000

In [13]:
query = '''

SELECT  books.book_id AS "Книга",
        "Количество обзоров",
        "Средний рейтинг"
FROM books
LEFT JOIN ( SELECT book_id,
            AVG(rating) AS "Средний рейтинг"
            FROM ratings
            GROUP BY book_id) AS SUB_4 ON books.book_id = SUB_4.book_id
LEFT JOIN ( SELECT book_id,
            COUNT(review_id) AS "Количество обзоров"
            FROM reviews
            GROUP BY book_id) AS SUBQ_5  ON books.book_id = SUBQ_5.book_id 

ORDER BY "Книга"
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

Книга  Количество обзоров  Средний рейтинг
0      1                   2         3.666667
1      2                   1         2.500000
2      3                   3         4.666667
3      4                   2         4.500000
4      5                   4         4.000000

### Вывод

Сформированным запросом мы получили датасет, где для каждой книги посчитана средняя оценка и количество обзоров.

### Задание 3.

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [10]:
query = '''
SELECT COUNT (books.book_id) AS "Количество книг",
       publisher AS Издательство
FROM books
LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher_id
ORDER BY "Количество книг" DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

Количество книг   Издательство
0               42  Penguin Books

### Вывод

Издательство __Penguin Books__ выпустило самое большое количество книг толщиной больше 50 страниц

### Задание 4.

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
query = '''
SELECT  authors.author AS "Автор",
        AVG("Средний рейтинг") AS "Средний рейтинг"     
FROM  authors
RIGHT JOIN (SELECT  books.author_id,
                    COUNT(rating_id) AS "Количество оценок",
                    AVG(rating) AS "Средний рейтинг"
            FROM books
            LEFT JOIN ratings ON books.book_id = ratings.book_id 
            GROUP BY books.book_id
            HAVING COUNT(rating_id) > 50) AS SUBQ_1 ON authors.author_id = SUBQ_1.author_id
GROUP BY authors.author_id
ORDER BY "Средний рейтинг" DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

Автор  Средний рейтинг
0  J.K. Rowling/Mary GrandPré         4.283844

### Вывод

При формировании запроса в подзаgросе SUBQ_1 оставили только те книги, у которых было 50 и более оценок. Всего таких книг оказалось 18.
После объединения таблиц и группировки данных был найдет автор с самой высокой средней оценкой - __J.K. Rowling/Mary GrandPré__


### Задание 5.

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
query = '''
SELECT AVG (SUBQ_3.count) AS "Cреднее количество обзоров"
FROM
    (SELECT  reviews.username,
            COUNT(review_id)
    FROM reviews
    INNER JOIN (SELECT  username,
                        COUNT (rating_id)
                FROM ratings
                GROUP BY username
                HAVING COUNT (rating_id) > 50) AS SUBQ_2 ON SUBQ_2.username = reviews.username
    GROUP BY reviews.username) AS SUBQ_3

'''
pd.io.sql.read_sql(query, con = engine)

Cреднее количество обзоров
0                   24.333333

### Вывод

При формировании запроса в подзапросе SUBQ_2 нашли пользователей, которые поставили более 50 оценок. Таких пользователей оказалось 6 человек. 
После объединения таблиц и группировки посчитали количество обзоров, который сделал каждый, а затем нашли среднее значение.
Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - более 24 обзоров.